Plotting script for bagfiles from robot

jonbinney opened the pull request:
In an attempt to figure out what is going wrong between the motion profile creation and the actual robot controller, this script plots both the command `JointTrajectory` and all positions `fanuc_j519_msgs/Command` messages.

Tested on this bag: https://drive.google.com/file/d/1d10WC443f0ofPPLrjbPbXhOg3MtmNpRO/view?usp=sharing

The goal trajectory looks smooth:
![goal_trajectory](https://user-images.githubusercontent.com/1538056/72944997-74c44b00-3d2f-11ea-9a9b-dbc814bcfb77.png)

But the commands sent to the robot show very high jerk:
![command_positions](https://user-images.githubusercontent.com/1538056/72945378-7e01e780-3d30-11ea-9ec1-442767533d7e.png)

The commands plot is zero for a while. The goal trajectory is sent to the joint_trajectory_controller at the the very end of the commanded positions graph. The initial commands result in very large jerk, and the robot immediately faults.

@gavanderhoorn - it looks like the commands being sent to the robot are large, so this is probably an issue with the trajectory i've made, or with the way joint_trajectory_controller is using it.

gavanderhoorn commented:
Or the driver implementation of course, I don't know.

I'll take a look at the bag.

gavanderhoorn commented:
I'm not entirely sure I'm doing it correctly, but a quick test with some bits borrowed from the `joint_trajectory_controller` and the `JointTrajectory` in the `.bag` you shared show that the first pt is getting dropped ("Requested sample time precedes trajectory start time").

Not sure exactly why just yet, but this could be the cause of the high jerk, as we're essentially skipping a sample at that point.

Ignoring the first pt (so starting at `0.008`) and printing the difference between the output of `sample(..)` (which is what the JTC uses internally) and what is in the `JointTrajectory` shows that it does appear to just return the state at the beginning of the segment (ie: no sampling by the JTC itself). So that is nice, as it means we don't have to change the JTC, as long as the trajectory is sufficiently dense.

The only anomaly is the first point, which has this (that's per-joint diff in radians):

```
0.000006702; 0.000003125; 0.000001343;-0.000000003;-0.000000004; 0.000000002
```

I believe things are rather sensitive in any case, as we're using an open-loop hw interface, which means that the JTC is just passing these pts on and the hw interface will also forward this immediately. If between the start of the generation/reception of the trajectory and the execution there has been (sufficient) movement of the robot, there will be a larger distance (in joint space) leading to even further increased jerk.

---

Edit: bounds checking on whether the first point (or any point really) is "before the current time" appears to be exclusive. See [here](https://github.com/ros-controls/ros_controllers/blob/b480d5f06568b31b8c68a54de8042e1d63d01534/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_msg_utils.h#L57).

I set the `stamp` of the traj msg to `0` and then start sampling at `0.0`, which of course means `0 < 0.0 => false`. This leads to the first segment getting dropped (as [this](https://github.com/ros-controls/ros_controllers/blob/b480d5f06568b31b8c68a54de8042e1d63d01534/joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h#L325) returns the first pt in the trajectory, but then [this](https://github.com/ros-controls/ros_controllers/blob/b480d5f06568b31b8c68a54de8042e1d63d01534/joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h#L332) skips to the next one).

Still not sure whether this is the cause of the problem, but getting closer.

gavanderhoorn commented:
Note btw that since we're only passing in positions, the JTC defaults to linear interpolation between pts (but again: I don't believe anything is getting interpolated).

jonbinney commented:
I got the arm to move by doing two things:
- added a bunch of JointTrajectoryPoints at the beginning of the trajectory with a position exactly equal to the robot's initial state. I did this to make sure it didn't matter if the JTC skipped the first point
- reduced the max jerk in the trajectory generation down to 30% of max.

With these changes, the arm successfully moved through the first portion of the trajectory, turning 90 degrees around joint 1 and tilting joint 2 down a bit, but as it is decelerating to zero at the waypoint it faults with a jerk  limit. I've plotted both the command positions and state positions, velocity, acceleration, and jerk for each joint separately:

![command_and_state](https://user-images.githubusercontent.com/1538056/73039830-4e281200-3e0c-11ea-98f9-d7cc0d89dace.png)

The problem happens just after t=10s on the plots. The trajectory itself looks very reasonable.

Here is the bagfile for this case: 

https://drive.google.com/file/d/1mLmD4pemV7jpSS_mEIoiDa_nRDb5jwpj/view?usp=sharing

I've pushed my updates to the plotting script. I've added the state message for the JTC to the bagfile so that we can see what the driver is trying to achieve.

gavanderhoorn commented:
It's probably correct, but that velocity for `joint_3` going in the opposite direction confuses me (there is link2-link3 linkage though, so might be caused by that).

It also doesn't seem to have the same magnitude.

gavanderhoorn commented:
> The trajectory itself looks very reasonable.

Well ..

![command](https://user-images.githubusercontent.com/4550046/73066442-3e5d0c00-3ea6-11ea-92a4-e9581509e62d.png)

The first derivative is even nastier.

Not sure what is going on yet.

---

Edit:

> I've added the state message for the JTC to the bagfile so that we can see what the driver is trying to achieve.

Default publish rate for `state` topic of the JTC is 50 Hz. Would be good to set that to 125 Hz as well (it's configured in `fanuc_m20ib_control/config/controllers.yaml`).

---

Edit 2: could you please also make a wireshark capture next time? The bag is missing at least one `RobotState` msg.

---

Edit 3:

> it faults with a jerk limit. 

have you made a controller backup? Did the error message tell you which joint was the cause?

gavanderhoorn commented:
PlotJuggler plots using the `header.stamp`, which is not synced with controller time (neither for `Command` nor `RobotState`). This causes artefacts, as the JTC is ran at `8ms` intervals precisely (in fact, `t` is never used, only `dt` and that is calculated based on the robot controller's `time_stamp`, which is always `8ms`).

Even though jitter in the ROS timestamps is low (stddev: `0.0001973132`) it's sufficient to cause these jerky plots.

Overwriting the timestamps gives this (this is for `joint_1`):

![joint_1](https://user-images.githubusercontent.com/4550046/73075450-3b6d1600-3ebc-11ea-8a6d-79d54f4162f0.png)

which is much better and comes close to your plots @jonbinney.

We should probably use `RobotState::time_stamp` for `RobotState` messages, but then the clocks need to be synchronised very well, or we'd need to compensate for drift.

gavanderhoorn commented:
Just taken another look at this, and I don't really understand why it's doing this.

For some reason, in `Command` pkt 4727, the velocity dips slightly for J1 (between `10.504` and `10.512`), causing a rather large deceleration (approx `-6600 deg/s^2`, and consequent jerk). That would seem to coincide with your description "the problem happens just after t=10s on the plots".

The top of the commanded position curve for J1 looks like this:

![j1_top_of_curve](https://user-images.githubusercontent.com/4550046/73091440-bba27400-3eda-11ea-921d-b6c7f351d8a1.png)

The up-to-down transition there is in 3 packets. It's unclear to me why the JTC would want to do that.

I haven't plotted the input trajectory around there yet. @jonbinney: could it be that the result of the stitching of the segments in the trajectory is somehow not continuous?

---

Edit: could you perhaps try a trajectory that does not decelerate and then accelerate again? Ie: something that moves in a single direction?

jonbinney commented:
I don't see any discontinuities, even when i take the actual generated FollowJointTrajectory goal, and compute the discrete derivatives in (i think) the same way that the fanuc does. I've updated the plotting script to plot the FollowJointTrajectory goal on the same axes as the commanded and state values from the driver. It's a little tricky because i don't see a good way of getting the exact start time of the trajectory from the JTC. Instead, I've added a commandline arg that lets the user guess and check the start time to get the graphs to line up.

The results are very interesting. It looks like the velocity of the commands sent by the driver are maxxing out at around 1.0 radians per second for some reason. Here are the position and velocity graphs for J1. Green is the goal trajectory given to the JTC. Blue is the commanded value the driver sends to the arm. Red is the state value red by the driver from the arm.

![Screenshot from 2020-01-24 13-28-07](https://user-images.githubusercontent.com/1538056/73105539-22f5ff00-3eae-11ea-9e9a-22374b0c994f.png)

jonbinney commented:
If you want to regenerate this graph yourself, the command is
```
ropsrun traj plot_bag <bag_directory>/j519_2020-01-23-17-58-22.bag --goal-start=6.82
```

jonbinney commented:
I reduced the max velocity in the demo script by about a factor of 3 and was able to move the arm without any faults! We're getting closer....

gavanderhoorn commented:
> The results are very interesting. It looks like the velocity of the commands sent by the driver are maxxing out at around 1.0 radians per second for some reason.

The driver uses `ros_control`'s saturation limits.

Could you please check your copy of ` fanuc_m20ib_support/config/joint_limits_m20ib25c.yaml` and/or your MoveIt configuration? I seem to remember @bhomberg changing those limits to avoid running into parameterisation issues in the past.

---

Edit: I'm sort of expecting the reduced joint limits to be the cause of what the plot shows.

> I reduced the max velocity in the demo script by about a factor of 3 and was able to move the arm without any faults! We're getting closer....

So the default velocity limit for J1 and J2 is `3.577925 rad/s`. Reducing the limits used by the scripts here by 3x would get them close to what the driver uses as limits if @bhomberg changed all limits to `1.0 rad/s`. That would lead to much less tracking error between the JTC and the real robot.

I'm sort of suspecting tracking error to be the real cause of the problems here: we're feeding the trajectory open-loop to the controller. The JTC has no way of dealing with tracking error, so it'll just push whatever `sample(t, ..)` returns. As tracking error increases, so does the chance of the JTC commanding something which falls outside of the limits of the real hw at that point in time.

I'd be really interested to see whether the limits have been changed by @bhomberg and whether that affects what we're seeing here at all.

Additional thought: it might not be a bad idea to switch to a [closed loop hw iface](https://github.com/ros-controls/ros_controllers/blob/b480d5f06568b31b8c68a54de8042e1d63d01534/joint_trajectory_controller/include/joint_trajectory_controller/hardware_interface_adapter.h#L134-L222) approach and/or configure/implement a tracking error checking mechanism.

gavanderhoorn commented:
@mahmoud-a-ali btw.

jonbinney commented:
Ah yes, we have set the velocity limits to 1.0 rad/sec using rosparams. I hadn't realized the driver was using those values.

For a jerk limited robot, capping velocity will always create problems, right? It makes a big acceleration and jerk spike. I'm not sure of a better solution though. We could modify the JTC to reject ahead of time if they violate limits.

jonbinney commented:
I'll keep an eye on the tracking error. I had assumed that the limits that StreamMotion puts on velocity/acceleration/jerk were chosen so that the robot should be able to accurately track any valid trajectory. If we have to switch to closed loop, that makes things more complicated. Not only would the initial trajectory need to be smooth, but the controller would need to be intelligent about not violating jerk limits.

jonbinney commented:
Success! I was able to move the robot at fairly high speeds. Velocity, acceleration all close to max speed for the robot.

jonbinney commented:
Here is a plot of the position/vel/acc/jerk for each joint, zoomed in at a wapoint where the robot decelerates and reverse direction.

![zoomed3](https://user-images.githubusercontent.com/1538056/73235259-8730f100-4142-11ea-83aa-1733754b8523.png)


Position, velocity, and acceleration look good. The robot state trails the commands by about 60ms, but that is about what we expect.

jonbinney commented:
@gavanderhoorn @mahmoud-a-ali i emailed you a video and bagfile of this test

gavanderhoorn commented:
@jonbinney wrote:
> Success! I was able to move the robot at fairly high speeds. Velocity, acceleration all close to max speed for the robot.

Very nice. This is indeed getting close to what we want.

Could you quantify "fairly high speeds"?

> The robot state trails the commands by about 60ms, but that is about what we expect.

Exactly. 5 to 6 ITPs, so about 40 to 50 ms.

gavanderhoorn commented:
> Ah yes, we have set the velocity limits to 1.0 rad/sec using rosparams. I hadn't realized the driver was using those values.

Well it's keeping an eye on the commands it receives from the controller(s) and uses saturation limiters to make sure we don't exceed any of those limits.

It's just velocity saturation though, but that was causing the velocity to be so nicely limited to 1 rad/s in your plots.

> For a jerk limited robot, capping velocity will always create problems, right?

A robot sensitive to jerk limits, yes. I would say all robots are jerk limited, but some have (quite extensive) filtering between input and execution.

> It makes a big acceleration and jerk spike. I'm not sure of a better solution though. We could modify the JTC to reject ahead of time if they violate limits.

I was thinking about something like this as well.

Right now "isValid(..)` in the JTC doesn't look at those properties of a trajectory, but we could cretainly extend it to do that. That would probably also be something upstream would be interested in.

> I'll keep an eye on the tracking error. I had assumed that the limits that StreamMotion puts on velocity/acceleration/jerk were chosen so that the robot should be able to accurately track any valid trajectory. 

Yes, that is indeed how the limits have been chosen afaik.

> If we have to switch to closed loop, that makes things more complicated. 

The problem in this case was on our side: we artificially capped the robot to 1 rad/s. The real hw is capable of more. The trajectory generator assumed the regular limits. This caused things to desync (in a way) leading to the JTC commanding a change in direction which resulted in a sign change of the tangent in a handful of `Command` pkts. That was not good.

> Not only would the initial trajectory need to be smooth, but the controller would need to be intelligent about not violating jerk limits.

yes, that would indeed be needed. But not necessarily too problematic. We don't need it to be very aggressive. It would just need to be able to cope with a slight difference in desired vs current state.

gavanderhoorn commented:
Just a note: getting this to work on Xenial is proving to be a bit of a pain.

gavanderhoorn commented on traj/package.xml:
```diff
@@ -9,4 +9,6 @@
 
   <buildtool_depend>catkin</buildtool_depend>
+
+  <exec_depend>python-sympy</exec_depend>
```
```suggestion
  <exec_depend>python-matplotlib</exec_depend>
  <exec_depend>python-numpy</exec_depend>
  <exec_depend>python-sympy</exec_depend>
```

jonbinney replied:
I manually made this change to the package.xml

gavanderhoorn commented on traj/scripts/demo:
```diff
@@ -66,16 +73,17 @@ max_jerks = np.deg2rad(np.array([
+joint_names = [joint_prefix + name for name in 'joint_1', 'joint_2', 'joint_3', 'joint_4', 'joint_5', 'joint_6']
 
 joint_states_sub = rospy.Subscriber('joint_states', JointState, joint_state_callback)
 while not rospy.is_shutdown() and initial_joint_states is None:
```
Would `rospy.wait_for_message('joint_states', sensor_msgs.msg.JointState)` not work?

jonbinney replied:
Whoa - i don't think i've ever used wait_for_message() before. That's great! I'll update the code to use it.

gavanderhoorn commented on traj/scripts/demo:
```diff
@@ -66,16 +73,17 @@ max_jerks = np.deg2rad(np.array([
 plot_trajectory = rospy.get_param('~plot', False)
+joint_prefix = rospy.get_param('~joint_prefix', '')
+
+joint_names = [joint_prefix + name for name in 'joint_1', 'joint_2', 'joint_3', 'joint_4', 'joint_5', 'joint_6']
```
Is this (ie: the inline list without the braces) Python `3.7` syntax? It errors out for me on `3.5`.

jonbinney replied:
It works in Python 2.7. I think that python is probably interpreting it as a tuple. It's unclear though; i'll add braces. (Are you using python 3.x with ROS?)

gavanderhoorn replied:
Hm, weird.

I'll try again.

> Are you using python 3.x with ROS?

Not in this workspace, but it's actually not that difficult. See [how to define ROS kinetic to use python3 instead of python2.7?](https://answers.ros.org/question/237613) fi (both Dirk's and my answer).

jonbinney replied:
Yeah, i know it's possible with a virtualenv. I have a bad habit of not using virtualenv though. I've been using clion even for python code lately though, and it has builtin virtualenv support, so I don't really have an excuse anymore.

jonbinney replied:
Brackets added

jonbinney commented:
I've made fixes based on your comments. By "high speed" I mean at least 70% of the limit (i dropped the limits down a bit from the actual robot limits to make sure we didn't get any errors from being right on the boundary). For joint 1, the robot reaches about 2.6 rad/sec velocity and 8.7 rad/sec**2 acceleration. Acceleration doesn't spend much (if any) time maxed out because of the way the limits relate to each other. The jerk limit is about 9x the acceleration limit, so it takes about 0.11 seconds or about 14 of the 8ms control cycles to reach max acceleration. But by that time the velocity is already so high that the acceleration needs to slow down.

Here is a plot of the entire trajectory:
![full_speed](https://user-images.githubusercontent.com/1538056/73302398-6fe91680-41c9-11ea-917e-350ededb2641.png)

gavanderhoorn commented:
joints 4, 5 and 6 are a bit noisy. But I guess that's caused by the very small motions. On my plots I can actually almost see the discretisation from the encoders. Would it be an idea to make 4, 5 and 6 go to `initial_joint_states[3:]` instead of `0.0, -0.3, 0.0`? Or is `0.0, -0.3, 0.0` a specific pose you need/want it to go to?

Something interesting: I executed a similar motion with our robot here, but then hard-coded TP.

Plotting the joint states for that motion it would appear acceleration and deceleration are not symmetric. It takes less time to accelerate than it does to decelerate (in other words: the robot takes a good long time to come to a complete stand-still). Velocity curves are lop-sided.

jonbinney commented:
The lopsided velocity curve is certainly interesting. Perhaps they are trying to avoid having the robot oscillate after reaching the goal? Do you have a bagfile or plot i could look at for that motion?

gavanderhoorn commented:
Not with me. It's only on the PC connected to the robot. I'll make it available tomorrow.

gavanderhoorn commented on traj/scripts/demo:
```diff
@@ -66,7 +66,7 @@ max_jerks = np.deg2rad(np.array([
     8000.0,
     16000.0,
-])) * 1.0
+]))
```
These are not really hacks, are they? :)

The scaling factor is essentially a way to slow down the trajectory. A bit primitive, but quite OK for now I'd say.

gavanderhoorn commented:
Is `demo` in this PR still supposed to work? I'm getting `import` errors:

```python-traceback
Traceback (most recent call last):
  File "./demo", line 15, in <module>
    import traj
  File "/home/user/catkin_ws/devel/lib/python2.7/dist-packages/traj/__init__.py", line 35, in <module>
    exec(__fh.read())
  File "<string>", line 6, in <module>
  File "/home/user/catkin_ws/src/trajectory_smoothing/traj/src/traj/trajectory.py", line 6, in <module>
    from seven_segment import fit_seven_segment
ImportError: No module named seven_segment
```

Substituting `seven_segment_type3` for `seven_segment` gets me in a really weird situation: the script runs, but hangs on `rospy.wait_for_message(..)`, doesn't print any of the `loginfo(..)`s, etc, etc.

Not really sure what is going on.

---

Edit: and finally this:

```python-traceback
Traceback (most recent call last):
  File "/home/user/catkin_ws/src/trajectory_smoothing/traj/scripts/demo", line 136, in <module>
    trajectory_jerk_function) = traj.trajectory_for_path(path, max_velocities, max_accelerations, max_jerks)
  File "/home/user/catkin_ws/src/trajectory_smoothing/traj/src/traj/trajectory.py", line 33, in trajectory_for_path
    v_max = project_limits_onto_s(max_velocities, fsegment)
  File "/home/user/catkin_ws/src/trajectory_smoothing/traj/src/traj/trajectory.py", line 9, in project_limits_onto_s
    slope = np.abs(np.array(diff(function)).astype(np.float64).flatten())
  File "/usr/lib/python2.7/dist-packages/sympy/core/function.py", line 1638, in diff
    return Derivative(f, *symbols, **kwargs)
  File "/usr/lib/python2.7/dist-packages/sympy/core/function.py", line 987, in __new__
    must be supplied to differentiate %s''' % expr))
ValueError: 
                    Since there is more than one variable in the
expression, the variable(s) of differentiation
must be supplied to differentiate Matrix([ [nan], [nan], [nan], [nan],
[nan], [nan]])
```

All on a Xenial+Kinetic machine btw.

@mahmoud-a-ali: are you still able to run the code in this PR?

mahmoud-a-ali commented:
@gavanderhoorn No, it does not work with me as well

jonbinney commented:
Hmmm... let me look into this.

jonbinney commented:
Rebasing on master fixes the import problem. The reason the script hangs is that it is waiting for a `joint_states` message to know where to start the trajectory from (I was testing this on the real robot). I'll add an option to use some default start configuration for the arm.

jonbinney commented:
Alright, I've rebased and fixed some issues caused by other code that got merged. You should now be able to run the demo script and see a plot:
```
rosrun traj demo
```

And if you add the fanuc repo to your workspace you can see a simulated robot follow the path in gazebo:
```
roslaunch traj sim_demo.launch
```

gavanderhoorn commented:
Thanks for the fixes, I'll test it later.

>  The reason the script hangs is that it is waiting for a `joint_states` message to know where to start the trajectory from (I was testing this on the real robot).

I'd realised that. As I had a real robot sitting right next to me (or at least: in its cage), I did not understand why `wait_for_message(..)` didn't return.

I checked the usual suspects though: node is definitely initialised, messages are being published, topic name correct. `rospy` shouldn't need any spinning. What is weird is that `rospy.log*(..)` is not printing for me, until I `ctrl+c` the script.

Does that happen for you as well?

jonbinney commented:
I didn't have that problem, but i  suspect that the script was hanging on the `wait_for_server()`, which came before any logging message. ctrl-c probable just knocked it out for that `wait_for_server()` call. 

With my newest changes, the script should run fine without a robot. It only looks for the FollowJointTrajectory action server if you set the `execute` parameter to tru.

jonbinney commented:
I believe that this is fixed; going to go ahead and merge.
